lib/routes/routeBackbeat.js

```diff
@@ -474,107 +474,98 @@ function putMetadata(request, response, bucketInfo, objMd, log, callback) {
         }
 
         let versionId = decodeVersionId(request.query);
-        let versioning = bucketInfo.isVersioningEnabled();
-        let isNull = false;
+        if (versionId instanceof Error) {
+            return callback(versionId);
+        }
 
-        if (versionId === 'null') {
-            isNull = true;
-            // Retrieve the null version id from the object metadata.
-            versionId = objMd && objMd.versionId;
-            if (!versionId) {
-                if (versioning) {
-                    // If the null version does not have a version id, it is a current null version.
-                    // To update the metadata of a current version, versioning is set to false.
+        return versioningPreprocessing(bucketName, bucketInfo, objectKey, omVal, log, (err, options) => {
```

Comment:
Normally, unless I'm mistaken, this function is never called for version-specific PUTs, it is responsible to process the versioning state for new version PUTs.

According to jsdoc:
> return versioning information for S3 to handle creation of new versions and manage deletion of old data and metadata

Indeed its side-effects are to either copy the null version to a new key or delete it, depending on circumstamces. I'm not sure this is what we want here.

I would rather only call this function if we are indeed doing a non-version-specific PUT to avoid side-effects like this (and do the right thing when indeed we are doing such operation).

Are the existing tests enough to cover this by the way?

Reply:
This existing test covers it https://github.com/scality/cloudserver/blob/bfa366cd272090084e41fd0b9d37e5a32df6c57f/tests/functional/raw-node/test/routes/routeBackbeat.js#L287

Reply:
I'm thinking of a test that
- updates a specific version of an object on a versioning-suspended bucket
- the same object also having a null version

I have the feeling in such case the null version would get deleted erroneously due to the use of `versioningPreprocessing` for a version-specific update.

I suggest to not use `versioningPreprocessing` when it is a version-specific update, because the function is not meant for this use.

By the way, is there a need to support regular versioned PUTs (or nonversioned) in the backbeat route? Maybe we don't need `versioningPreprocessing` after all?

Reply:
At least for 7.x branches the API will fail your request with a 409 for version suspended buckets, I'd need to check the behavior in 8.x.
7.x branches don't need to support non-versioned, but 8.x branches do. (based on the test comments)
So instead of `versioningPreprocessing` should I just implement the behavior in the handler directly? or perhaps break it out into `versioning.js`?

Reply:
@tmacro yes, implementing the logic directly in the handler should be fine (I'm not aware of duplicate code in Cloudserver but maybe it exists). Also good to use helpers in `versioning.js` if that makes sense and is not too specific to backbeat.

Reply:
We had lots of issues with null versions and various use cases on 8.x, I think now we should be good in most cases in there : so you may want to check how it was done there indeed, and possibly just backport some changes/commits...

Reply:
@jonathan-gramain I;ve pushed some changes